docker/docker_client.go

```diff
@@ -221,10 +221,87 @@ func CheckAuth(ctx context.Context, sCtx *types.SystemContext, username, passwor
 	}
 }
 
+// Data holds the output of both the v2 and v1 endpoints
+type Data struct {
+	// Repositories holds the results returned by the /v2/_catalog endpoint
+	Repositories []string `json:"repositories"`
+	// Results holds the results returned by the /v1/search endpoint
+	Results []Results `json:"results"`
+}
+
+// Results holds the information of each matching image returned by the v1 endpoint
+type Results struct {
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	StarCount   int    `json:"star_count"`
+	IsTrusted   bool   `json:"is_trusted"`
+	IsAutomated bool   `json:"is_automated"`
+	IsOfficial  bool   `json:"is_official"`
+}
+
+// SearchRegistries queries a list of registries for a matching image
+func SearchRegistries(ctx context.Context, sCtx *types.SystemContext, registries []string, image, limit string) (map[string][]Results, error) {
+	searchResults := make(map[string][]Results)
+	for _, reg := range registries {
+		url := reg
+		if strings.Contains(reg, "docker") {
+			url = "index.docker.io"
+		}
+
+		logrus.Debugf("pinging v2 endpoint\n")
+		data, err := getRegistryData(ctx, sCtx, url, image, limit)
+		if err != nil {
+			fmt.Printf("couldn't search registry %q: %v", reg, err)
+			continue
+		}
+
+		if len(data.Results) > 0 {
+			searchResults[reg] = data.Results
+			continue
+		}
+
+		results := []Results{}
+		for _, repo := range data.Repositories {
+			if strings.Contains(repo, image) {
+				res := Results{
+					Name: repo,
+				}
+				results = append(results, res)
+			}
+		}
+		searchResults[reg] = results
+	}
+	return searchResults, nil
+}
+
+// getREgistryData talks to either the v2 or v1 endpoint to get the results of the search query
+func getRegistryData(ctx context.Context, sCtx *types.SystemContext, registry, image, limit string) (*Data, error) {
+	client, err := newDockerClientWithDetails(sCtx, registry, "", "", "", nil, "")
```

Comment:
So from what I read online, looks like no form of authentication is required to get a response from the search endpoint, that is why I have not incorporated credentials in here. Do you thing I should?

Reply:
If we don’t know of a registry that requires authentication, it makes sense not to bother implementing it (and have it completely untested).

Still, I’d prefer the external API to allow for authentication to be added in the future (see the `index.docker.io` mapping discussion.)

Reply:
So you are saying that `SearchRegistry` should have username and password as parameters and if in future we find out that registries need authentication to search we can use that? I don't think it is something we need in podman yet, but I could definitely add it to the search registry function.

Reply:
No, I was thinking that the username/password could in the future come from `config.GetAuthentication`; that already allows the caller to override the username/password (which should be rare) via `types.SystemContext.DockerAuthConfig`.

So, no need to add unimplemented username/password parameters to the API now; but the hypothetical `config.GetAuthentication` lookup would be easier if the `registry` parameter for Docker Hub were `docker.io`, with the `index.docker.io` mapping performed somewhere inside this code, instead of being user-visible.

Reply:
ah okay, I get it now. Yeah I can do the mapping in here instead.

Reply:
fixed.